packages/flutter/lib/src/widgets/image_filter.dart

```diff
@@ -46,32 +65,75 @@ class ImageFiltered extends SingleChildRenderObjectWidget {
   }
 }
 
-class _ImageFilterRenderObject extends RenderProxyBox {
-  _ImageFilterRenderObject(this._imageFilter);
+/// Applies an [ImageFilter] to its child.
+class RenderImageFiltered extends RenderProxyBox {
+  /// Creates a [RenderObject] that applies an [ImageFilter] to its child.
+  ///
+  /// One of the [imageFilter] or [imageFilterCallback] should not be null here
+  /// or whenever the [paint] method is invoked.
+  /// The [imageFilter] will be used if it is not null, otherwise the filter
+  /// will be obtained from the [imageFilterCallback].
+  RenderImageFiltered({
+    required ImageFilter? imageFilter,
+    required ImageFilterCallback? imageFilterCallback,
+  })
+      : assert(imageFilter != null || imageFilterCallback != null),
+        _imageFilter = imageFilter,
+        _imageFilterCallback = imageFilterCallback;
 
-  ImageFilter get imageFilter => _imageFilter;
-  ImageFilter _imageFilter;
-  set imageFilter(ImageFilter value) {
-    assert(value != null);
+  /// The [ImageFilter] to apply to this child, or null if the filter will be supplied
+  /// instead by the [imageFilterCallback].
+  ///
+  /// If the [imageFilter] is set to null here, then either it or the [imageFilterCallback]
+  /// should be set to a non-null value before [paint] is called.
+  ImageFilter? get imageFilter => _imageFilter;
+  ImageFilter? _imageFilter;
+  set imageFilter(ImageFilter? value) {
     if (value != _imageFilter) {
       _imageFilter = value;
       markNeedsPaint();
     }
   }
 
+  /// Called to create the [ImageFilter] to apply to the child if the [imageFilter]
+  /// property is null.
+  ///
+  /// The image filter callback is called with the current bounds of the child so that
+  /// it can customize the filter to the size and location of the child.
+  ///
+  /// If the [imageFilterCallback] is set to null here, then either it or the [imageFilter]
+  /// should be set to a non-null value before [paint] is called.
+  ImageFilterCallback? get imageFilterCallback => _imageFilterCallback;
+  ImageFilterCallback? _imageFilterCallback;
+  set imageFilterCallback(ImageFilterCallback? value) {
+    if (value != _imageFilterCallback) {
+      _imageFilterCallback = value;
+      markNeedsPaint();
+    }
+  }
+
   @override
   bool get alwaysNeedsCompositing => child != null;
 
+  ImageFilter? _previousFilter;
+
   @override
   void paint(PaintingContext context, Offset offset) {
-    assert(imageFilter != null);
+    assert(imageFilter != null || imageFilterCallback != null);
+    ImageFilter newFilter = imageFilter ?? imageFilterCallback!(offset & size);
```

Comment:
do we need to call it again if the rect hasn't changed?

Reply:
Presumably we won't be asked to repaint if the rect hasn't changed? ShaderMask doesn't seem to care about that and just always calls the delegate.